java-frontend/src/main/java/org/sonar/java/se/checks/DivisionByZeroCheck.java

```diff
@@ -0,0 +1,258 @@
+    }
+
+    public static ZeroConstraint zero(Tree syntaxNode) {
+      return new ZeroConstraint(syntaxNode, Status.ZERO);
+    }
+
+    public static ZeroConstraint nonZero(Tree syntaxNode) {
+      return new ZeroConstraint(syntaxNode, Status.NON_ZERO);
+    }
+
+    @Override
+    public ZeroConstraint inverse() {
+      Tree syntaxNode = syntaxNode();
+      return isZero() ? nonZero(syntaxNode) : zero(syntaxNode);
+    }
+
+    private boolean isZero() {
+      return hasStatus(Status.ZERO);
+    }
+
+    @Override
+    public boolean isInversible() {
+      return isZero();
+    }
+
+    @Override
+    public String toString() {
+      // For debugging purpose
+      return "" + (isZero() ? Status.ZERO : Status.NON_ZERO);
+    }
+  }
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    PreStatementVisitor visitor = new PreStatementVisitor(this, context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  private static class PreStatementVisitor extends CheckerTreeNodeVisitor {
+    private final SECheck check;
+    private final ConstraintManager constraintManager;
+    private final CheckerContext context;
+
+    PreStatementVisitor(SECheck check, CheckerContext context) {
+      super(context.getState());
+      this.check = check;
+      this.context = context;
+      this.constraintManager = context.getConstraintManager();
+    }
+
+    private void addZeroConstraint(boolean isZero) {
+      constraintManager.setValueFactory((id, node) -> new ZeroSymbolicValue(id, isZero));
+    }
+
+    private static boolean isZero(ProgramState programState, SymbolicValue symbolicValue) {
+      return programState.getConstraintWithStatus(symbolicValue, Status.ZERO) != null;
+    }
+
+    private void reportIssue(ExpressionTree expression, Tree.Kind kind) {
+      String operation = (kind == Tree.Kind.REMAINDER || kind == Tree.Kind.REMAINDER_ASSIGNMENT) ? "modulation" : "division";
+      String expressionName = expression.is(Tree.Kind.IDENTIFIER) ? ("'" + ((IdentifierTree) expression).name() + "'") : "this expression";
+      context.reportIssue(expression, check, "Make sure " + expressionName + " can't be zero before doing this " + operation + ".");
+
+      // interrupt exploration
+      programState = null;
+    }
+
+    @Override
+    public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+      List<SymbolicValue> symbolicValues = programState.peekValues(2);
+      boolean varIsZero = isZero(programState, symbolicValues.get(0));
+      boolean exprIsZero = isZero(programState, symbolicValues.get(1));
+
+      switch (tree.kind()) {
+        case MULTIPLY_ASSIGNMENT:
+          addZeroConstraint(exprIsZero);
+          break;
+        case MINUS_ASSIGNMENT:
+        case PLUS_ASSIGNMENT:
+          addZeroConstraint(varIsZero);
+          break;
+        case DIVIDE_ASSIGNMENT:
+        case REMAINDER_ASSIGNMENT:
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this switch case number of lines from 6 to at most 5, for example by extracting code into methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1151)